File: book-ax-web/generate-icons.py
#!/usr/bin/env python3
"""
Generate all required icon formats for Book.ax Web App
- favicon.ico (16x16, 32x32, 48x48)
- apple-touch-icon.png (180x180)
- logo.png (512x512)
- og-image.jpg (1200x630)
"""

import os
from pathlib import Path

try:
    from PIL import Image, ImageDraw, ImageFont, ImageFilter
    import cairosvg
except ImportError:
    print("❌ Required packages not installed!")
    print("\n🔧 Run: pip3 install pillow cairosvg")
    print("   On macOS: brew install cairo pango gdk-pixbuf libffi")
    exit(1)

# Paths
PUBLIC_DIR = Path(__file__).parent / "public"
FAVICON_SVG = PUBLIC_DIR / "favicon.svg"
LOGO_SVG = PUBLIC_DIR / "logo.svg"

def svg_to_png(svg_path, png_path, width, height=None):
    """Convert SVG to PNG with specified dimensions"""
    if height is None:
        height = width
    
    print(f"  📐 {svg_path.name} → {png_path.name} ({width}x{height}px)")
    
    cairosvg.svg2png(
        url=str(svg_path),
        write_to=str(png_path),
        output_width=width,
        output_height=height
    )

def create_favicon_ico():
    """Create multi-resolution favicon.ico"""
    print("\n🎨 Creating favicon.ico...")
    
    # Generate PNG versions at different sizes
    sizes = [16, 32, 48, 64, 128, 256]
    temp_pngs = []
    
    for size in sizes:
        temp_png = PUBLIC_DIR / f"favicon-{size}.png"
        svg_to_png(FAVICON_SVG, temp_png, size)
        temp_pngs.append(Image.open(temp_png))
    
    # Save as multi-resolution ICO
    ico_path = PUBLIC_DIR / "favicon.ico"
    temp_pngs[0].save(
        ico_path,
        format='ICO',
        sizes=[(s, s) for s in sizes],
        append_images=temp_pngs[1:]
    )
    
    # Cleanup temp files
    for size in sizes:
        (PUBLIC_DIR / f"favicon-{size}.png").unlink()
    
    print(f"  ✅ favicon.ico created (multi-resolution)")

def create_apple_touch_icon():
    """Create Apple Touch Icon (180x180px)"""
    print("\n🍎 Creating apple-touch-icon.png...")
    
    png_path = PUBLIC_DIR / "apple-touch-icon.png"
    svg_to_png(FAVICON_SVG, png_path, 180)
    
    print(f"  ✅ apple-touch-icon.png created (180x180px)")

def create_logo_png():
    """Create logo.png for structured data (512x512px)"""
    print("\n📦 Creating logo.png...")
    
    png_path = PUBLIC_DIR / "logo.png"
    svg_to_png(LOGO_SVG, png_path, 512, 512)
    
    print(f"  ✅ logo.png created (512x512px)")

def create_og_image():
    """Create OpenGraph image for social sharing (1200x630px)"""
    print("\n🌐 Creating og-image.jpg...")
    
    # Create gradient background
    width, height = 1200, 630
    img = Image.new('RGB', (width, height))
    draw = ImageDraw.Draw(img)
    
    # Gradient from purple to violet
    for y in range(height):
        r = int(139 + (124 - 139) * (y / height))  # 139 → 124
        g = int(92 + (58 - 92) * (y / height))      # 92 → 58
        b = int(246 + (237 - 246) * (y / height))   # 246 → 237
        draw.rectangle([(0, y), (width, y+1)], fill=(r, g, b))
    
    # Load and paste logo (centered)
    try:
        # Convert logo SVG to PNG temporarily
        temp_logo = PUBLIC_DIR / "temp-logo-og.png"
        svg_to_png(LOGO_SVG, temp_logo, 600, 180)
        
        logo = Image.open(temp_logo)
        
        # Center logo
        logo_x = (width - logo.width) // 2
        logo_y = (height - logo.height) // 2
        
        # Paste with transparency
        if logo.mode == 'RGBA':
            img.paste(logo, (logo_x, logo_y), logo)
        else:
            img.paste(logo, (logo_x, logo_y))
        
        temp_logo.unlink()
    except Exception as e:
        print(f"  ⚠️  Could not add logo to OG image: {e}")
    
    # Add text
    try:
        # Try to use system fonts
        try:
            font_large = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 80)
            font_small = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 40)
        except:
            font_large = ImageFont.load_default()
            font_small = ImageFont.load_default()
        
        # Tagline
        tagline = "Find Your Perfect Stay"
        bbox = draw.textbbox((0, 0), tagline, font=font_small)
        text_width = bbox[2] - bbox[0]
        text_x = (width - text_width) // 2
        
        # Add shadow
        draw.text((text_x + 2, 480 + 2), tagline, fill=(0, 0, 0, 100), font=font_small)
        draw.text((text_x, 480), tagline, fill='white', font=font_small)
        
    except Exception as e:
        print(f"  ⚠️  Could not add text: {e}")
    
    # Save as JPG with high quality
    og_path = PUBLIC_DIR / "og-image.jpg"
    img.save(og_path, 'JPEG', quality=95, optimize=True)
    
    print(f"  ✅ og-image.jpg created (1200x630px)")

def main():
    print("=" * 60)
    print("🎨 Book.ax Icon Generator")
    print("=" * 60)
    
    if not FAVICON_SVG.exists():
        print(f"❌ Error: {FAVICON_SVG} not found!")
        return
    
    if not LOGO_SVG.exists():
        print(f"❌ Error: {LOGO_SVG} not found!")
        return
    
    # Generate all icons
    create_favicon_ico()
    create_apple_touch_icon()
    create_logo_png()
    create_og_image()
    
    print("\n" + "=" * 60)
    print("✅ All icons generated successfully!")
    print("=" * 60)
    print("\n📂 Generated files:")
    print("   • favicon.ico       (multi-resolution)")
    print("   • apple-touch-icon.png (180x180px)")
    print("   • logo.png          (512x512px)")
    print("   • og-image.jpg      (1200x630px)")
    print()

if __name__ == "__main__":
    main()

File: book-ax-web/generate-icons.js
#!/usr/bin/env node
/**
 * Generate all required icon formats for Book.ax Web App
 * Uses sharp for image processing (no Python dependencies needed)
 */

const fs = require('fs').promises;
const path = require('path');

async function checkSharp() {
  try {
    const sharp = require('sharp');
    return sharp;
  } catch (error) {
    console.log('❌ Sharp not installed!');
    console.log('\n🔧 Run: npm install sharp --save-dev\n');
    process.exit(1);
  }
}

async function generateIcons() {
  const sharp = await checkSharp();
  
  console.log('=' + '='.repeat(59));
  console.log('🎨 Book.ax Icon Generator (Node.js)');
  console.log('=' + '='.repeat(59));
  
  const publicDir = path.join(__dirname, 'public');
  const faviconSvg = path.join(publicDir, 'favicon.svg');
  const logoSvg = path.join(publicDir, 'logo.svg');
  
  // Check if SVG files exist
  try {
    await fs.access(faviconSvg);
    await fs.access(logoSvg);
  } catch (error) {
    console.log('❌ SVG files not found!');
    return;
  }
  
  console.log('\n🎨 Creating favicon.ico...');
  // Generate favicon.ico (using 32x32 as base)
  await sharp(faviconSvg)
    .resize(32, 32)
    .png()
    .toFile(path.join(publicDir, 'favicon-32.png'));
  
  // For ICO, we'll use PNG for now (browsers support it)
  await sharp(faviconSvg)
    .resize(32, 32)
    .toFormat('png')
    .toFile(path.join(publicDir, 'favicon.ico'));
  
  console.log('  ✅ favicon.ico created (32x32px)');
  
  console.log('\n🍎 Creating apple-touch-icon.png...');
  await sharp(faviconSvg)
    .resize(180, 180)
    .png()
    .toFile(path.join(publicDir, 'apple-touch-icon.png'));
  console.log('  ✅ apple-touch-icon.png created (180x180px)');
  
  console.log('\n📦 Creating logo.png...');
  await sharp(logoSvg)
    .resize(512, 512)
    .png()
    .toFile(path.join(publicDir, 'logo.png'));
  console.log('  ✅ logo.png created (512x512px)');
  
  console.log('\n🌐 Creating og-image.jpg...');
  // Create gradient background with logo
  const width = 1200;
  const height = 630;
  
  // Create purple gradient SVG
  const gradientSvg = `
    <svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <linearGradient id="grad" x1="0%" y1="0%" x2="0%" y2="100%">
          <stop offset="0%" style="stop-color:#8b5cf6;stop-opacity:1" />
          <stop offset="100%" style="stop-color:#7c3aed;stop-opacity:1" />
        </linearGradient>
      </defs>
      <rect width="${width}" height="${height}" fill="url(#grad)"/>
      <text x="50%" y="55%" font-family="Arial, sans-serif" font-size="120" font-weight="700" fill="white" text-anchor="middle" dominant-baseline="middle">Book.ax</text>
      <text x="50%" y="75%" font-family="Arial, sans-serif" font-size="48" font-weight="400" fill="white" text-anchor="middle" opacity="0.9">Find Your Perfect Stay</text>
    </svg>
  `;
  
  await sharp(Buffer.from(gradientSvg))
    .jpeg({ quality: 95 })
    .toFile(path.join(publicDir, 'og-image.jpg'));
  console.log('  ✅ og-image.jpg created (1200x630px)');
  
  // Cleanup temp files
  try {
    await fs.unlink(path.join(publicDir, 'favicon-32.png'));
  } catch (e) {}
  
  console.log('\n' + '=' + '='.repeat(59));
  console.log('✅ All icons generated successfully!');
  console.log('=' + '='.repeat(59));
  console.log('\n📂 Generated files:');
  console.log('   • favicon.ico       (32x32px)');
  console.log('   • apple-touch-icon.png (180x180px)');
  console.log('   • logo.png          (512x512px)');
  console.log('   • og-image.jpg      (1200x630px)');
  console.log();
}

generateIcons().catch(error => {
  console.error('❌ Error:', error.message);
  process.exit(1);
});
